agent/dockerclient/dockerapi/docker_events_buffer.go

```diff
@@ -51,15 +52,25 @@ func NewInfiniteBuffer() *InfiniteBuffer {
 }
 
 // StartListening starts reading from the input channel and writes to the buffer
-// TODO: wire in ctx to stop listening
-func (buffer *InfiniteBuffer) StartListening(events chan *docker.APIEvents) {
-	for event := range events {
-		go buffer.CopyEvents(event)
+// When context is cancelled, stop listening
+func (buffer *InfiniteBuffer) StartListening(ctx context.Context, eventChan <-chan events.Message) {
+	for {
+		select {
+		// If context is cancelled, return
+		case <-ctx.Done():
+			return
+		default:
+			for event := range eventChan {
```

Comment:
You already have a for loop outside, so you don't need to do a for loop here, the go routine will be stuck here forever and you won't get a chance to return when context is done. 

Also, when the context is done, we need to make sure all the events in the channel are read so that we don't miss anything,  below is my suggestion:

```
func (buffer *InfiniteBuffer) StartListening(ctx context.Context, eventChan <-chan events.Message) {
   for {
      select {
      // If context is cancelled, return
      case <-ctx.Done():
         select {
         // Ensure all events are read in the channel before we return
         case event <- eventChan:
            go func(j events.Message) {
               go buffer.CopyEvents(&j)
            }(event)
         default:
            return
         }
      case event <- eventChan:
         go func(j events.Message) {
            go buffer.CopyEvents(&j)
         }(event)
      }
   }
}
```